authentik/lib/avatars.py

```diff
@@ -181,12 +162,44 @@ def avatar_mode_generated(user: "User", mode: str) -> str | None:
 def avatar_mode_url(user: "User", mode: str) -> str | None:
     """Format url"""
     mail_hash = md5(user.email.lower().encode("utf-8"), usedforsecurity=False).hexdigest()  # nosec
-    return mode % {
+
+    formatted_url = mode % {
         "username": user.username,
         "mail_hash": mail_hash,
         "upn": user.attributes.get("upn", ""),
     }
 
+    hostname = urlparse(formatted_url).hostname
+    cache_key_hostname_available = f"goauthentik.io/lib/avatars/{hostname}/available"
+
+    if not cache.get(cache_key_hostname_available, True):
+        return None
+
+    cache_key_image_url = f"goauthentik.io/lib/avatars/{hostname}/{mail_hash}"
+
+    if cache.has_key(cache_key_image_url):
+        cache.touch(cache_key_image_url)
+        return cache.get(cache_key_image_url)
+
+    try:
+        res = get_http_session().head(formatted_url, timeout=5, allow_redirects=True)
+
+        if res.status_code == HttpResponseNotFound.status_code:
+            cache.set(cache_key_image_url, None)
+            return None
+        if not res.headers.get("Content-Type", "").startswith("image/"):
+            cache.set(cache_key_image_url, None)
+            return None
+        res.raise_for_status()
+    except (Timeout, ConnectionError, HTTPError):
+        cache.set(cache_key_hostname_available, False, timeout=GRAVATAR_STATUS_TTL_SECONDS)
+        return None
+    except RequestException:
+        return formatted_url
+
+    cache.set(cache_key_image_url, formatted_url, timeout=GRAVATAR_STATUS_TTL_SECONDS)
```

Comment:
I would probably rename those constants with AVATAR_URL as a prefix instead of GRAVATAR